### PR TITLE
UI Test Framework

### PR DESCRIPTION
**Pull request by GDRDABARERA**

### CODE REVIEW

File: tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/SignInTest.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+package org.wso2.is.portal.user.test.ui;
+
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.htmlunit.HtmlUnitDriver;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+import org.wso2.carbon.iam.userportal.actionobject.LoginPageAction;
+
+
```
Comment (ayshsandu) [inline tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/SignInTest.java]: Remove unnecessary new lines

File: tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/UsernameRecoveryTest.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import org.openqa.selenium.htmlunit.HtmlUnitDriver;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+import org.wso2.carbon.iam.userportal.actionobject.UsernameRecoveryPageAction;
+import org.wso2.carbon.identity.mgt.connector.Attribute;
+
+import java.util.ArrayList;
+import java.util.List;
+
+
```
Comment (ayshsandu) [inline tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/UsernameRecoveryTest.java]: Remove unnecessary new lines

File: portal/apps/org.wso2.carbon.iam.user-portal.app/src/test/java/org/wso2/carbon/iam/userportal/actionobject/LoginPageAction.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+package org.wso2.carbon.iam.userportal.actionobject;
+
+import org.openqa.selenium.WebDriver;
+import org.wso2.carbon.iam.userportal.pageobject.LoginPage;
+
+/**
+ * Action class for login.
+ */
+public class LoginPageAction {
```
Comment (ayshsandu) [inline portal/apps/org.wso2.carbon.iam.user-portal.app/src/test/java/org/wso2/carbon/iam/userportal/actionobject/LoginPageAction.java]: Rename to LoginAction

File: portal/apps/org.wso2.carbon.iam.user-portal.app/pom.xml [hunk 4]
```diff
@@ -59,6 +59,30 @@
             <groupId>org.wso2.carbon.identity.mgt</groupId>
             <artifactId>org.wso2.carbon.identity.recovery</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.seleniumhq.selenium</groupId>
+            <artifactId>selenium-api</artifactId>
+            <version>2.40.0</version>
```
Comment (omindu) [inline portal/apps/org.wso2.carbon.iam.user-portal.app/pom.xml]: Can't we add the versions as properties in parent pom. Add the dependency to DependencyManagement section in the parent.

File: tests/commons/pom.xml [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (omindu) [inline tests/commons/pom.xml]: Missing licence header.

File: portal/apps/org.wso2.carbon.iam.user-portal.app/src/test/java/org/wso2/carbon/iam/userportal/actionobject/LoginPageAction.java (hunk 3 above)
Comment (GDRDABARERA) [inline portal/apps/org.wso2.carbon.iam.user-portal.app/src/test/java/org/wso2/carbon/iam/userportal/actionobject/LoginPageAction.java]: All the actions of a single page is written in a single action class so this class is LoginPageAction

File: tests/pom.xml [hunk 6]
```diff
@@ -31,6 +31,7 @@
 
     <modules>
         <module>osgi-tests</module>
+	    <module>ui-tests</module>
```
Comment (omindu) [inline tests/pom.xml]: Formatting issue.

File: tests/commons/pom.xml [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+    <parent>
+        <artifactId>user-portal-tests</artifactId>
+        <groupId>org.wso2.is</groupId>
+        <version>6.0.0-SNAPSHOT</version>
+    </parent>
+
+    <groupId>org.wso2.is</groupId>
+    <artifactId>commons</artifactId>
+    <version>6.0.0-SNAPSHOT</version>
```
Comment (omindu) [inline tests/commons/pom.xml]: No need to specify the groupId version. Will be inherited from the parent pom

File: tests/ui-tests/user-portal/pom.xml [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    <dependencies>
+        <dependency>
+            <groupId>org.testng</groupId>
+            <artifactId>testng</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.testng</groupId>
+            <artifactId>testng</artifactId>
+            <version>6.3.1</version>
```
Comment (omindu) [inline tests/ui-tests/user-portal/pom.xml]: Use the testng version property in parent pom.

File: tests/ui-tests/user-portal/pom.xml [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.is</groupId>
+            <artifactId>org.wso2.carbon.iam.user-portal.feature</artifactId>
+            <version>6.0.0-SNAPSHOT</version>
```
Comment (omindu) [inline tests/ui-tests/user-portal/pom.xml]: Do we need the version here ?

File: tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/SignInTest.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    public void testLogin() throws Exception {
+        String username = "admin";
+        String password = "admin";
+        loginPageAction.login(driver, username, password);
+        Assert.assertEquals(driver.getCurrentUrl(), adminPage);
+        driver.close();
+        driver = new HtmlUnitDriver();
+        driver.get(loginPage);
+        testClickUsernameRecovery(driver);
+
```
Comment (omindu) [inline tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/SignInTest.java]: Unnecessary new line

File: tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/SignInTest.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+public class SignInTest {
+
+    private static LoginPageAction loginPageAction = new LoginPageAction();
+    private static WebDriver driver = new HtmlUnitDriver();
+    private static String loginPage = "https://localhost:9292/user-portal/login";
+    private static String adminPage = "https://localhost:9292/user-portal/";
+    private static String usernameRecoveryPage = "https://localhost:9292/user-portal/recovery/username";
+    private static String passwordRecoveryPage = "https://localhost:9292/user-portal/recovery/password";
+
```
Comment (omindu) [inline tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/SignInTest.java]: Unnecessary new line

File: tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/SignInTest.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.htmlunit.HtmlUnitDriver;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+import org.wso2.carbon.iam.userportal.actionobject.LoginPageAction;
+
+
+
+public class SignInTest {
```
Comment (omindu) [inline tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/SignInTest.java]: Add class level comment.

File: tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/UsernameRecoveryTest.java [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+    @Test(groups = "usernameRecoveryTest")
+    public void loadUsernameRecoveryPage() throws Exception {
+        driver.get(usernameRecoveryPage);
+
+    }
+
+    @Test(groups = "usernameRecoveryTest", dependsOnMethods = "loadUsernameRecoveryPage")
+    public void testUssernameRecovery() throws Exception {
+        String[] claims = {"givenname", "lastname", "email"};
```
Comment (omindu) [inline tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/UsernameRecoveryTest.java]: Can use Map<String, String>.

File: tests/ui-tests/user-portal/pom.xml [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+            <groupId>org.seleniumhq.selenium</groupId>
+            <artifactId>selenium-htmlunit-driver</artifactId>
+            <version>${org.seleniumhq.selenium.version}</version>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
```
Comment (omindu) [inline tests/ui-tests/user-portal/pom.xml]: This won't be needed as it's already configured in wso2 parent pom

File: tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/SignInTest.java [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    public void loadLoginPage() throws Exception {
+        driver.get(loginPage);
+    }
+
+    @Test(groups = "signInTest", dependsOnMethods = "loadLoginPage")
+    public void testLogin() throws Exception {
+        String username = "admin";
+        String password = "admin";
+        loginPageAction.login(driver, username, password);
+        Assert.assertEquals(driver.getCurrentUrl(), adminPage);
```
Comment (omindu) [inline tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/SignInTest.java]: Include a message mentioning why the test failed.

File: tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/SignInTest.java [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    @Test(groups = "signInTest", dependsOnMethods = "loadLoginPage")
+    public void testClickUsernameRecovery(WebDriver driver) throws Exception {
+        loginPageAction.clickForgetUsername(driver);
+        Assert.assertEquals(driver.getCurrentUrl(), usernameRecoveryPage);
+        driver.close();
+        driver = new HtmlUnitDriver();
+        driver.get(loginPage);
+        testClickPasswordRecovery(driver);
```
Comment (omindu) [inline tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/SignInTest.java]: Is this necessary ?

File: portal/apps/org.wso2.carbon.iam.user-portal.app/src/test/java/org/wso2/carbon/iam/userportal/pageobject/LoginPage.java [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+
+/**
+ * Represents login page
+ */
+public class LoginPage {
+    private static WebElement element = null;
+
+    public static WebElement txtbx_UserName(WebDriver driver) {
```
Comment (omindu) [inline portal/apps/org.wso2.carbon.iam.user-portal.app/src/test/java/org/wso2/carbon/iam/userportal/pageobject/LoginPage.java]: Any reason why `_` is used in method names ? We normally don't do this.

File: tests/ui-tests/pom.xml [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+  ~ under the License.
+  -->
+
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>user-portal-tests</artifactId>
+        <groupId>org.wso2.is</groupId>
+        <version>6.0.0-SNAPSHOT</version>
```
Comment (omindu) [inline tests/ui-tests/pom.xml]: Missing parent relative path.

File: tests/ui-tests/user-portal/pom.xml [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+  ~ under the License.
+  -->
+
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>product-is-tests-ui</artifactId>
+        <groupId>org.wso2.is</groupId>
+        <version>6.0.0-SNAPSHOT</version>
```
Comment (omindu) [inline tests/ui-tests/user-portal/pom.xml]: Missing parent relative path.

File: pom.xml [hunk 20]
```diff
@@ -949,6 +961,13 @@
 
         <javax.ws.rs.import.version.range>[2.0.0, 3.0.0)</javax.ws.rs.import.version.range>
 
+        <!-- Selenium -->
+        <org.seleniumhq.selenium.version>2.40.0</org.seleniumhq.selenium.version>
```
Comment (dharshanaw) [inline pom.xml]: Update selinium version to latest released version 3.1.0

File: tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/SignInTest.java [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+ */
+public class SignInTest {
+
+    private static LoginPageAction loginPageAction = new LoginPageAction();
+    private static WebDriver driver;
+    private static String loginPage = "https://localhost:9292/user-portal/login";
+    private static String adminPage = "https://localhost:9292/user-portal/";
+    private static String usernameRecoveryPage = "https://localhost:9292/user-portal/recovery/username";
+    private static String passwordRecoveryPage = "https://localhost:9292/user-portal/recovery/password";
+
```
Comment (dharshanaw) [inline tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/SignInTest.java]: Dont hardcode localhost. Refer host name and port by a configuration file. You can use old framework engine and automation.xml for this

File: tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/UsernameRecoveryTest.java [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+/**
+ * UI Tests for Username Recovery.
+ */
+public class UsernameRecoveryTest {
+
+    private static UsernameRecoveryPageAction usernameRecoveryPageAction
+            = new UsernameRecoveryPageAction();
+    private static WebDriver driver;
+    private static String  loginPage = "https://localhost:9292/user-portal/login";
+    private static String usernameRecoveryPage = "https://localhost:9292/user-portal/recovery/username";
```
Comment (dharshanaw) [inline tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/UsernameRecoveryTest.java]: Remove hardcoded host and port

File: tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/SignInTest.java [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+        driver.get(loginPage);
+        driver.close();
+    }
+
+    @Test(groups = "signInTest", dependsOnMethods = "loadLoginPage")
+    public void testLogin() throws Exception {
+        driver = new HtmlUnitDriver();
+        driver.get(loginPage);
+        String username = "admin";
+        String password = "admin";
```
Comment (dharshanaw) [inline tests/ui-tests/user-portal/src/test/java/org/wso2/is/portal/user/test/ui/SignInTest.java]: Dont hard code admin username and password

File: tests/ui-tests/user-portal/src/test/resources/testng.xml [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+  ~ KIND, either express or implied.  See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
+
+<suite name="WSO2-Identity-Server-user-portal-ui-Test-Suite">
+    <test name="wso2-identity-server-user-portal-ui" preserve-order="true" parallel="false">
+        <!--Commenting these tests, because in order to run these there should be a IS 6.0 server running. -->
```
Comment (dharshanaw) [inline tests/ui-tests/user-portal/src/test/resources/testng.xml]: Remove commented out elements

Comment (dharshanaw) [inline tests/ui-tests/user-portal/src/test/resources/testng.xml]: Add test here

File: tests/ui-tests/user-portal/pom.xml [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
```
Comment (dharshanaw) [inline tests/ui-tests/user-portal/pom.xml]: Add surefire plugin

File: tests/ui-tests/user-portal/src/test/resources/testng.xml (hunk 24 above)
Comment (GDRDABARERA) [inline tests/ui-tests/user-portal/src/test/resources/testng.xml]: Added they are commented till kernel upgrade finishes.

File: tests/ui-tests/user-portal/pom.xml (hunk 25 above)
Comment (GDRDABARERA) [inline tests/ui-tests/user-portal/pom.xml]: done

**Comment by GDRDABARERA** [comment]
Closed this with a new PR https://github.com/wso2/product-is/pull/1175

**Comment by GDRDABARERA** [comment]
Closed this with this new PR https://github.com/wso2/product-is/pull/1175

**Pull request closed without merging**